libopeniscsiusr/idbm.c

```diff
@@ -0,0 +1,699 @@
+	check_password_param(discovery.sendtargets.auth.password_in);
+	check_password_param(discovery.slp.auth.password);
+	check_password_param(discovery.slp.auth.password_in);
+	check_password_param(host.auth.password);
+	check_password_param(host.auth.password_in);
+
+out:
+	return rc;
+}
+
+/*
+ * from linux kernel
+ */
+static char *strstrip(char *s)
+{
+	size_t size;
+	char *end;
+
+	size = strlen(s);
+	if (!size)
+		return s;
+
+	end = s + size - 1;
+	while (end >= s && isspace(*end))
+		end--;
+	*(end + 1) = '\0';
+
+	while (*s && isspace(*s))
+		s++;
+
+	return s;
+}
+
+static int _idbm_recs_read(struct iscsi_context *ctx, struct idbm_rec *recs,
+			   const char *conf_path)
+{
+	int rc = LIBISCSI_OK;
+	char name[NAME_MAXVAL];
+	char value[VALUE_MAXVAL];
+	char *line = NULL;
+	char *nl = NULL;
+	char buffer[2048];
+	int line_number = 0;
+	int c = 0;
+	int i = 0;
+	FILE *f = NULL;
+	int errno_save = 0;
+	char strerr_buff[_STRERR_BUFF_LEN];
+
+	assert(ctx != NULL);
+	assert(recs != NULL);
+	assert(conf_path != NULL);
+
+	f = fopen(conf_path, "r");
+	errno_save = errno;
+	if (!f) {
+		_error(ctx, "Failed to open %s using read mode: %d %s",
+		       conf_path, errno_save,
+		       _strerror(errno_save, strerr_buff));
+		rc = LIBISCSI_ERR_IDBM;
+		goto out;
+	}
+
+	_info(ctx, "Parsing iSCSI interface configuration %s", conf_path);
+	/* process the config file */
+	do {
+		line = fgets(buffer, sizeof (buffer), f);
+		line_number++;
+		if (!line)
+			continue;
+		if (strlen(line) == 0)
+			continue;
+
+		nl = line + strlen(line) - 1;
+		if (*nl != '\n') {
+			_warn(ctx, "Config file %s line %d too long.",
+			      conf_path, line_number);
+			continue;
+		}
+
+		line = strstrip(line);
+		/* process any non-empty, non-comment lines */
+		if (!*line || *line == '\0' || *line ==  '\n' || *line == '#')
+			continue;
+
+		/* parse name */
+		i=0; nl = line; *name = 0;
+		while (*nl && !isspace(c = *nl) && *nl != '=') {
+			*(name+i) = *nl; i++; nl++;
+		}
+		if (!*nl) {
+			_warn(ctx, "config file %s line %d do not has value",
+			      conf_path, line_number);
+			continue;
+		}
+		*(name+i)=0; nl++;
+		/* skip after-name traling spaces */
+		while (*nl && isspace(c = *nl)) nl++;
+		if (*nl && *nl != '=') {
+			_warn(ctx, "config file %s line %d has not '=' "
+			      "separator", conf_path, line_number);
+			continue;
+		}
+		/* skip '=' sepa */
+		nl++;
+		/* skip after-sepa traling spaces */
+		while (*nl && isspace(c = *nl)) nl++;
+		if (!*nl) {
+			_warn(ctx, "config file %s line %d do not has value",
+			      conf_path, line_number);
+			continue;
+		}
+		/* parse value */
+		i=0; *value = 0;
+		while (*nl) {
+			*(value+i) = *nl; i++; nl++;
+		}
+		*(value+i) = 0;
+
+		rc = _idbm_rec_update_param(ctx, recs, name, value,
+					    line_number);
+		if (rc == LIBISCSI_ERR_INVAL) {
+			_error(ctx, "config file %s invalid.", conf_path);
+			goto out;
+		} else if (rc != LIBISCSI_OK)
+		      goto out;
+	} while (line);
```

Comment:
Please supply a sample config file that this is intending to parse, thanks!

Reply:
In the repo `./etc/iface.example` would do.